src/core/bbr.c

```diff
@@ -0,0 +1,1327 @@
+        //
+        // We are pacing, so split the congestion window into chunks which are
+        // spread out over the RTT. Calculate the current send allowance (chunk
+        // size) as the time since the last send times the pacing rate (CWND / RTT).
+        //
+        if (Bbr->BbrState == BBR_STATE_STARTUP) {
+            SendAllowance = (uint32_t)CXPLAT_MAX(
+                BandwidthEst * Bbr->PacingGain * TimeSinceLastSend / BBR_UNIT,
+                CongestionWindow * Bbr->PacingGain / BBR_UNIT - Bbr->BytesInFlight);
+        } else {
+            SendAllowance = (uint32_t)(BandwidthEst * Bbr->PacingGain * TimeSinceLastSend / BBR_UNIT);
+        }
+
+        if (SendAllowance > CongestionWindow - Bbr->BytesInFlight) {
+            SendAllowance = CongestionWindow - Bbr->BytesInFlight;
+        }
+
+        if (SendAllowance > (CongestionWindow >> 2)) {
+            SendAllowance = CongestionWindow >> 2; // Don't send more than a quarter of the current window.
+        }
+    }
+    return SendAllowance;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+void
+BbrCongestionControlHandleAckInProbeBw(
+    _In_ QUIC_CONGESTION_CONTROL* Cc,
+    _In_ uint64_t AckTime,
+    _In_ uint64_t PrevInflightBytes,
+    _In_ BOOLEAN HasLoss
+    )
+{
+    QUIC_CONGESTION_CONTROL_BBR* Bbr = &Cc->Bbr;
+
+    BOOLEAN ShouldAdvancePacingGainCycle =
+        CxPlatTimeDiff64(AckTime, Bbr->CycleStart) > BbrCongestionControlGetMinRtt(Cc);
+
+    if (Bbr->PacingGain > BBR_UNIT && !HasLoss &&
+        PrevInflightBytes < BbrCongestionControlCalculateTargetCwnd(Cc, Bbr->PacingGain)) {
+        //
+        // pacingGain_ > BBR_UNIT means BBR is probeing bandwidth. So we should let inflight bytes reach the target.
+        //
+        ShouldAdvancePacingGainCycle = FALSE;
+    }
+
+    if (Bbr->PacingGain < BBR_UNIT) {
+        uint64_t TargetCwnd = BbrCongestionControlCalculateTargetCwnd(Cc, BBR_UNIT);
+        if (Bbr->BytesInFlight <= TargetCwnd) {
+            //
+            // pacingGain_ < BBR_UNIT means BBR is draining the network queue. If inflight bytes is below
+            // the target, then it's done.
+            //
+            ShouldAdvancePacingGainCycle = TRUE;
+        }
+    }
+
+    if (ShouldAdvancePacingGainCycle) {
+        Bbr->PacingCycleIndex = (Bbr->PacingCycleIndex + 1) % kNumOfCycles;
+        Bbr->CycleStart = AckTime;
+        Bbr->PacingGain = kPacingGainCycles[Bbr->PacingCycleIndex];
+    }
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+BOOLEAN
+BbrCongestionControlDetectShouldExitStartup(
+    _In_ QUIC_CONGESTION_CONTROL* Cc
+    )
+{
+    return Cc->Bbr.BbrState == BBR_STATE_STARTUP && Cc->Bbr.BtlbwFound;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+void
+BbrCongestionControlDetectBottleneckBandwidth(
+    _In_ QUIC_CONGESTION_CONTROL* Cc,
+    _In_ BOOLEAN IsAppLimited
+    )
+{
+    QUIC_CONGESTION_CONTROL_BBR* Bbr = &Cc->Bbr;
+
+    if (Bbr->BtlbwFound) {
+        return;
+    }
+
+    if (IsAppLimited) {
+        return;
```

Comment:
Just don't call BbrCongestionControlDetectBottleneckBandwidth from BbrCongestionControlOnDataAcknowledged if LastAckedPacketAppLimited is TRUE.

In fact, please just move the code from this "helper" function into BbrCongestionControlOnDataAcknowledged.